### PR TITLE
[Enhancement] parquet late materialize round-by-round

### PR DESCRIPTION
**Pull request by zombee0**
Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 62 / 68 (91.18%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/runtime/types.cpp|8|9|88.89%|[386]|
|:large_blue_circle:|src/formats/parquet/group_reader.cpp|51|56|91.07%|[209, 210, 223, 310, 412]|
|:large_blue_circle:|src/exec/hdfs_scanner_parquet.cpp|3|3|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

File: be/src/formats/parquet/group_reader.cpp [hunk 1]
```diff
@@ -286,6 +280,63 @@ Status GroupReader::_read_range(const std::vector<int>& read_columns, const Rang
...
+        if (std::find(_dict_column_indices.begin(), _dict_column_indices.end(), col_idx) !=
+            _dict_column_indices.end()) {
+            SCOPED_RAW_TIMER(&_param.stats->expr_filter_ns);
+            SCOPED_RAW_TIMER(&_param.stats->group_dict_filter_ns);
+            for (const auto& sub_field_path : _dict_column_sub_field_paths[col_idx]) {
+                _column_readers[slot_id]->filter_dict_column((*chunk)->get_column_by_slot_id(slot_id), filter,
+                                                             sub_field_path, 0);
+                hit_count = SIMD::count_nonzero(*filter);
+                if (hit_count == 0) {
+                    goto end;
```
Comment (DorianZheng) [inline be/src/formats/parquet/group_reader.cpp:310]: use break instead

Comment (zombee0) [reply be/src/formats/parquet/group_reader.cpp:310]: we need to break out from two layers for-loop, goto will be more cleaner

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 74 / 82 (90.24%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/formats/parquet/round_by_round_ctx.h|4|5|80.00%|[27]|
|:large_blue_circle:|src/formats/parquet/group_reader.cpp|44|50|88.00%|[208, 209, 222, 303, 304, 399]|
|:large_blue_circle:|src/runtime/types.cpp|8|9|88.89%|[386]|
|:large_blue_circle:|src/exec/hdfs_scanner_parquet.cpp|3|3|100.00%|[]|
|:large_blue_circle:|src/formats/parquet/round_by_round_ctx.cpp|15|15|100.00%|[]|

File: be/src/formats/parquet/round_by_round_ctx.h [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+// limitations under the License.
+
+#pragma once
+
+#include <unordered_map>
+#include <vector>
+
+namespace starrocks::parquet {
+
+class RoundByRoundCtx {
```
Comment (DorianZheng) [inline be/src/formats/parquet/round_by_round_ctx.h:22]: `class ColumnOrder`

File: be/src/formats/parquet/round_by_round_ctx.h [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+public:
+    RoundByRoundCtx(std::vector<int> col_idxs, size_t all_cost, std::unordered_map<int, size_t> col_cost)
+            : _column_indices(col_idxs), _min_round_cost(all_cost), _column_cost_map(std::move(col_cost)) {}
+
+    ~RoundByRoundCtx() = default;
+
+    const std::vector<int>& get_round_order();
```
Comment (DorianZheng) [inline be/src/formats/parquet/round_by_round_ctx.h:29]: `get()`

File: be/src/formats/parquet/group_reader.cpp [hunk 4]
```diff
@@ -286,6 +279,51 @@ Status GroupReader::_read_range(const std::vector<int>& read_columns, const Rang
...
+        if (std::find(_dict_column_indices.begin(), _dict_column_indices.end(), col_idx) !=
+            _dict_column_indices.end()) {
+            SCOPED_RAW_TIMER(&_param.stats->expr_filter_ns);
+            SCOPED_RAW_TIMER(&_param.stats->group_dict_filter_ns);
+            for (const auto& sub_field_path : _dict_column_sub_field_paths[col_idx]) {
+                _column_readers[slot_id]->filter_dict_column((*chunk)->get_column_by_slot_id(slot_id), filter,
+                                                             sub_field_path, 0);
+                hit_count = SIMD::count_nonzero(*filter);
+                if (hit_count == 0) {
+                    _round_by_round_ctx->update_ctx(round_cost);
```
Comment (DorianZheng) [inline be/src/formats/parquet/group_reader.cpp:303]: Duplicate with LOC:323~324

File: be/src/formats/parquet/group_reader.cpp [hunk 5]
```diff
@@ -286,6 +279,51 @@ Status GroupReader::_read_range(const std::vector<int>& read_columns, const Rang
...
 }
 
+StatusOr<size_t> GroupReader::_read_range_round_by_round(const Range<uint64_t>& range, Filter* filter,
+                                                         ChunkPtr* chunk) {
+    const std::vector<int>& round_order = _round_by_round_ctx->get_round_order();
+    size_t round_cost = 0;
+    size_t hit_count = 0;
+    for (int col_idx : round_order) {
+        auto& column = _param.read_cols[col_idx];
+        round_cost += _round_by_round_ctx->get_column_cost(col_idx);
```
Comment (DorianZheng) [inline be/src/formats/parquet/group_reader.cpp:289]: should we get cost from column reader which can record cpu cost

File: be/src/formats/parquet/group_reader.h [hunk 6]
```diff
@@ -165,9 +171,13 @@ class GroupReader {
     // columns(index) use as dict filter column
     std::vector<int> _dict_column_indices;
     std::unordered_map<int, std::vector<std::vector<std::string>>> _dict_column_sub_field_paths;
+    std::unordered_map<SlotId, std::vector<ExprContext*>> _left_conjunct_ctxs_by_slot_id;
```
Comment (DorianZheng) [inline be/src/formats/parquet/group_reader.h:174]: Extract all conjuncts into a class leads to high cohesion

File: be/src/formats/parquet/group_reader.cpp [hunk 7]
```diff
@@ -286,6 +279,51 @@ Status GroupReader::_read_range(const std::vector<int>& read_columns, const Rang
...
+                _column_readers[slot_id]->read_range(range, filter, (*chunk)->get_column_by_slot_id(slot_id).get()));
+
+        if (std::find(_dict_column_indices.begin(), _dict_column_indices.end(), col_idx) !=
+            _dict_column_indices.end()) {
+            SCOPED_RAW_TIMER(&_param.stats->expr_filter_ns);
+            SCOPED_RAW_TIMER(&_param.stats->group_dict_filter_ns);
+            for (const auto& sub_field_path : _dict_column_sub_field_paths[col_idx]) {
+                _column_readers[slot_id]->filter_dict_column((*chunk)->get_column_by_slot_id(slot_id), filter,
+                                                             sub_field_path, 0);
+                hit_count = SIMD::count_nonzero(*filter);
```
Comment (DorianZheng) [inline be/src/formats/parquet/group_reader.cpp:301]: can we get hit count from `filter_dict_column`

File: be/src/formats/parquet/group_reader.h (hunk 6 above)
Comment (DorianZheng) [inline be/src/formats/parquet/group_reader.h:174]: What does left conjunct mean?

**Review by DorianZheng — APPROVED** [review]
left comments, rest LGTM.

File: be/src/formats/parquet/group_reader.cpp (hunk 4 above)
Comment (zombee0) [reply be/src/formats/parquet/group_reader.cpp:303]: Will use defer

File: be/src/formats/parquet/group_reader.cpp (hunk 5 above)
Comment (zombee0) [reply be/src/formats/parquet/group_reader.cpp:289]: We left comment to optimize

File: be/src/formats/parquet/round_by_round_ctx.h (hunk 2 above)
Comment (zombee0) [reply be/src/formats/parquet/round_by_round_ctx.h:22]: ColumnReadOrder will make sence

File: be/src/formats/parquet/group_reader.cpp (hunk 7 above)
Comment (zombee0) [reply be/src/formats/parquet/group_reader.cpp:301]: the predicate evaluate does not return hit count, we need calculate it.

File: be/src/formats/parquet/group_reader.h (hunk 6 above)
Comment (zombee0) [reply be/src/formats/parquet/group_reader.h:174]: It means _left_no_dict_filter_conjunct_by_slot_it, dict filter will push down and rewrite, others left, will rename _left_no_dict_filter_conjuncts_by_slot

**Review by dirtysalt — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 74 / 82 (90.24%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/formats/parquet/round_by_round_ctx.h|4|5|80.00%|[27]|
|:large_blue_circle:|src/formats/parquet/group_reader.cpp|44|50|88.00%|[208, 209, 222, 303, 304, 399]|
|:large_blue_circle:|src/runtime/types.cpp|8|9|88.89%|[386]|
|:large_blue_circle:|src/exec/hdfs_scanner_parquet.cpp|3|3|100.00%|[]|
|:large_blue_circle:|src/formats/parquet/round_by_round_ctx.cpp|15|15|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 77 / 85 (90.59%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/formats/parquet/column_read_order_ctx.h|4|5|80.00%|[27]|
|:large_blue_circle:|src/formats/parquet/group_reader.cpp|47|53|88.68%|[208, 209, 223, 224, 300, 395]|
|:large_blue_circle:|src/runtime/types.cpp|8|9|88.89%|[386]|
|:large_blue_circle:|src/exec/hdfs_scanner_parquet.cpp|3|3|100.00%|[]|
|:large_blue_circle:|src/formats/parquet/column_read_order_ctx.cpp|15|15|100.00%|[]|

**Review by Smith-Cruise — APPROVED** [review]

File: be/src/runtime/types.cpp [hunk 8]
```diff
@@ -381,6 +381,21 @@ int TypeDescriptor::get_slot_size() const {
     return -1;
 }
 
+size_t TypeDescriptor::get_flat_size() const {
```
Comment (Youngwb) [inline be/src/runtime/types.cpp:384]: get_children_size()?

**Review by Youngwb — APPROVED** [review]

**Pull request merged**